### ISSUE TITLE
Can not get negative risk

### ISSUE THREAD

**Issue by WangXin93** [kiryor/NNPUlearning#5]
Hi,

I tried to implement this algorithm in ``pytorch``, but found the risk can not go negative. Due to this, the **nnPU** and **uPU** has the same training loss history, can you help me figure out what's wrong with my implementation?

Here is my pytorch code:

```python
class PULoss(nn.Module):
    """Loss function for PU learning."""
    def __init__(self, prior, loss=(lambda x: torch.sigmoid(-x)), gamma=1, beta=0, nnPU=True):
        super(PULoss, self).__init__()
        if not 0 < prior < 1:
            raise NotImplementedError("The class prior should be in (0, 1)")
        self.prior = prior
        self.gamma = gamma
        self.beta = beta
        self.loss_func = loss
        self.nnPU = nnPU
        self.positive = 1
        self.unlabeled = -1

    def forward(self, x, t):
        t = t[:, None]
        positive, unlabeled = (t==self.positive).float(), (t==self.unlabeled).float()
        n_positive, n_unlabeled = max(1., positive.sum().item()), max(1., unlabeled.sum().item())
        y_positive = self.loss_func(x)
        y_unlabeled = self.loss_func(x)
        positive_risk = torch.sum(self.prior * positive / n_positive * y_positive)
        negative_risk = torch.sum((unlabeled / n_unlabeled - self.prior * positive / n_positive) * y_unlabeled)
        objective = positive_risk + negative_risk
        if self.nnPU:
            if negative_risk.item() < -self.beta:  # No negative occasion in my experiment here !!!
                objective = positive_risk -self.beta
                self.x_out = -self.gamma * negative_risk
            else:
                self.x_out = objective
        else:
            self.x_out = objective
        self.loss = objective
        return self.loss, self.out
```
![train_loss_history](https://user-images.githubusercontent.com/19688994/50592977-6d869100-0ed1-11e9-82b0-016a05b0405e.png)

I don't know if it is correct to use auto-gradient function here, I can sure the forward pass is the same as ``pu_loss.py`` in this repository.

Thanks a lot!

**Comment by fYYw** [comment]
`y_unlabeled = self.loss_func(x)` should it be `-x`?

**Comment by kiryor** [comment]
I'm not super familiar with pytorch, but the answer from @fYYw seems correct.
